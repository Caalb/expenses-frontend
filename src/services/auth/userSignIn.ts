import type { AxiosResponse } from 'axios'
import api from '../api'

export interface UserSignInResponse {
  token: string
}

export const userSignIn = async (
  email: string,
  password: string,
): Promise<AxiosResponse<UserSignInResponse>> => {
  const response = await api.post<UserSignInResponse>('/auth/signin', { email, password })

  return response
}
