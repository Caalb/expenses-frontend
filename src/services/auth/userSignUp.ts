import type { AxiosResponse } from 'axios'
import api from '../api'

export interface UserSignUpResponse {
  token: string
}

export const userSignUp = async (
  name: string,
  email: string,
  password: string,
): Promise<AxiosResponse<UserSignUpResponse>> => {
  const response = await api.post<UserSignUpResponse>('/auth/signup', { name, email, password })

  return response
}
