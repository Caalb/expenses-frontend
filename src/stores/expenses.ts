import { createExpense as createExpenseApi } from '@/services/expenses/createExpense'
import { fetchExpenses as fetchExpensesApi } from '@/services/expenses/fetchExpenses'
import type { Expense } from '@/types/expense'
import { defineStore } from 'pinia'
import { ref } from 'vue'

export const useExpensesStore = defineStore('expenses', () => {
  const expenses = ref<Expense[]>([])

  const fetchExpenses = async () => {
    try {
      const response = await fetchExpensesApi()

      expenses.value = response

      return response
    } catch (error) {
      return error
    }
  }

  const createExpense = async (expense: Omit<Expense, 'id'>) => {
    try {
      const response = await createExpenseApi(expense)
      expenses.value = [response.data, ...expenses.value]

      return response
    } catch (error) {
      return error
    }
  }

  return {
    expenses,
    fetchExpenses,
    createExpense,
  }
})
