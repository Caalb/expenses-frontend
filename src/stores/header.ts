import { defineStore } from 'pinia'
import { ref } from 'vue'

export const useHeaderStore = defineStore('header', () => {
  const activeSubmenu = ref<string | null>(null)

  const setActiveSubmenu = (submenu: string) => {
    activeSubmenu.value = activeSubmenu.value === submenu ? null : submenu
  }

  const clearSubmenu = () => {
    activeSubmenu.value = null
  }

  return {
    activeSubmenu,
    setActiveSubmenu,
    clearSubmenu,
  }
})
