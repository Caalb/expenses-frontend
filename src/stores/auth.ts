import { defineStore } from 'pinia'
import { userSignIn } from '../services/auth/userSignIn'
import type { AxiosResponse } from 'axios'
import { userSignUp } from '@/services/auth/userSignUp'

export const useAuthStore = defineStore('auth', () => {
  const handleSignIn = async (
    email: string,
    password: string,
  ): Promise<AxiosResponse | unknown> => {
    try {
      const response = await userSignIn(email, password)
      const { token } = response.data

      localStorage.setItem('token', token)
      return response
    } catch (error: unknown) {
      return error
    }
  }

  const handleSignUp = async (data: Record<string, string>): Promise<AxiosResponse | unknown> => {
    try {
      const response = await userSignUp(data.name, data.email, data.password)
      const { token } = response.data

      localStorage.setItem('token', token)
      return response
    } catch (error: unknown) {
      return error
    }
  }

  return {
    handleSignIn,
    handleSignUp,
  }
})
