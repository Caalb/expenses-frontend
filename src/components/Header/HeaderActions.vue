<template>
  <div class="row items-center q-gutter-sm">
    <QBtn
      round
      flat
      class="q-ml-sm"
    >
      <QAvatar size="32px">
        <img src="https://e1.pngegg.com/pngimages/896/660/png-clipart-pikachu-bymika-pikachu-illustration-thumbnail.png" />
      </QAvatar>

      <QTooltip>Perfil</QTooltip>

      <QMenu auto-close>
        <QList style="min-width: 200px">
          <QItem
            clickable
            @click="userLogout"
          >
            <QItemSection avatar>
              <QIcon name="logout" />
            </QItemSection>

            <QItemSection>Sair</QItemSection>
          </QItem>
        </QList>
      </QMenu>
    </QBtn>
  </div>
</template>

<script setup lang="ts">
import {
  QBtn,
  QAvatar,
  QMenu,
  QList,
  QItem,
  QItemSection,
  QIcon,
  QTooltip
} from 'quasar'
import { useAuthStore } from '@/stores/auth'
import { useQuasar } from 'quasar'
import { useRouter } from 'vue-router'

const router = useRouter()
const $q = useQuasar()
const { handleLogout } = useAuthStore()

const userLogout = () => {
  handleLogout()
  router.push({ name: 'login' })
  $q.notify({
    color: 'negative',
    message: 'Sess√£o encerrada com sucesso!',
    icon: 'check',
  })
}
</script>
