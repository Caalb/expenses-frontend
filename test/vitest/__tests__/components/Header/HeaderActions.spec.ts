import HeaderActions from '@/components/Header/HeaderActions.vue'
import router from '@/router'
import { createTestingPinia } from '@pinia/testing'
import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import { Notify } from 'quasar'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import type { ComponentPublicInstance } from 'vue'
import { useRouter } from 'vue-router'
import { useAuthStore } from '@/stores/auth'

interface HeaderActionsInstance extends ComponentPublicInstance {
  userLogout: () => void
}

installQuasarPlugin({ plugins: { Notify } })

vi.mock('vue-router')

const createWrapper = () =>
  mount(HeaderActions, {
    global: {
      plugins: [
        createTestingPinia({
          createSpy: vi.fn,
        }),
      ],
      mocks: {
        $router: {
          push: vi.fn(),
        },
      },
    },
  }) as VueWrapper<HeaderActionsInstance>

describe('HeaderActions', () => {
  let wrapper: VueWrapper<HeaderActionsInstance>
  let store: ReturnType<typeof useAuthStore>

  vi.mocked(useRouter).mockReturnValue({
    ...router,
    push: vi.fn(),
  })

  beforeEach(() => {
    wrapper = createWrapper()
    store = useAuthStore()
  })

  it('should match snapshot', () => {
    const wrapper = createWrapper()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should display the user profile avatar', () => {
    expect(wrapper.find('img').attributes('src')).toBe(
      'https://e1.pngegg.com/pngimages/896/660/png-clipart-pikachu-bymika-pikachu-illustration-thumbnail.png',
    )
  })

  it('should call userLogout', async () => {
    const spy = vi.spyOn(wrapper.vm.$q, 'notify')
    wrapper.vm.userLogout()

    expect(useRouter().push).toHaveBeenCalledWith({ name: 'login' })
    expect(spy).toHaveBeenCalled()
    expect(store.handleLogout).toHaveBeenCalled()
  })
})
