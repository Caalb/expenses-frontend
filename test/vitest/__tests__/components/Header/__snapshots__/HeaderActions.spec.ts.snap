// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HeaderActions > should match snapshot 1`] = `"<div class="row items-center q-gutter-sm"><button class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round q-btn--actionable q-focusable q-hoverable q-ml-sm" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><div class="q-avatar" style="font-size: 32px;"><div class="q-avatar__content row flex-center overflow-hidden"><img src="https://e1.pngegg.com/pngimages/896/660/png-clipart-pikachu-bymika-pikachu-illustration-thumbnail.png"></div></div><!----><!----></span></button></div>"`;
